addons/godot-next/nodes/behaviors.gd

```diff
@@ -0,0 +1,173 @@
+	var list := []
+	list.append(PropertyInfo.new_dictionary("_behaviors", 0, "", PROPERTY_USAGE_STORAGE).to_dict())
+	list += _behaviors.get_collection_property_list()
+	return list
+
+func _ready() -> void:
+	if not Engine.editor_hint:
+		var behaviors = _behaviors.get_data()
+		for behavior in behaviors.values():
+			_initialize_behavior(behavior)
+		
+		for behavior in _callbacks["_ready"]:
+			behavior._ready()
+
+func _process(delta: float) -> void:
+	for behavior in _callbacks["_process"]:
+		if behavior.enabled:
+			behavior._process(delta)
+
+func _physics_process(delta: float) -> void:
+	for behavior in _callbacks["_physics_process"]:
+		if behavior.enabled:
+			behavior._process(delta)
+
+func _input(event: InputEvent) -> void:
+	for behavior in _callbacks["_input"]:
+		if behavior.enabled:
+			behavior._input(event)
+
+func _unhandled_input(event: InputEvent) -> void:
+	for behavior in _callbacks["_unhandled_input"]:
+		if behavior.enabled:
+			behavior._unhandled_input(event)
+
+func _unhandled_key_input(event: InputEventKey) -> void:
+	for behavior in _callbacks["_unhandled_key_input"]:
+		if behavior.enabled:
+			behavior._process(event)
+
+##### VIRTUALS #####
+
+func _add_element(p_variable: String, p_script: Script) -> void:
+	var collection: ResourceCollection = get(p_variable)
+	collection.add_element(p_script)
+
+##### PUBLIC METHODS #####
+
+func add_behavior(p_type: Script) -> Behavior:
+	var behaviors = _behaviors.get_data()
+	
+	var ct = ClassType.new(p_type)
+	if not ct.type_is(Behavior):
+		return null
+	if has_behavior(p_type):
+		return get_behavior(p_type)
+		
+	var behavior: Behavior = p_type.new()
+	var behavior_name: String = ct.get_script_class()
+	
+	behaviors[behavior_name] = behavior
+	_initialize_behavior(behavior)
+	
+	return behavior
+
+func get_behavior(p_type: Script) -> Behavior:
+	var behaviors = _behaviors.get_data()
+	
+	var ct = ClassType.new(p_type)
+	var type_name: String = ct.get_script_class()
+	if behaviors.has(type_name):
+		return behaviors[type_name]
+	return null
+
+func has_behavior(p_type: Script) -> bool:
+	var behaviors = _behaviors.get_data()
+	return behaviors.has(ClassType.new(p_type).get_script_class())
+
+func remove_behavior(p_type: Script) -> bool:
+	var behaviors = _behaviors.get_data()
+	
+	var type_name := ClassType.new(p_type).get_script_class()
+	if not behaviors.has(type_name):
+		return false
+	var behavior = behaviors[type_name]
+	#warning-ignore:return_value_discarded
+	behaviors.erase(type_name)
+	_remove_from_callbacks(behavior)
+	behavior.free()
+	return true
+
+##### PRIVATE METHODS #####
+
+func _initialize_behavior(p_behavior: Behavior) -> void:
+	p_behavior.awake(self)
+	#warning-ignore:return_value_discarded
+	p_behavior.connect("script_changed", self, "_refresh_callbacks", [p_behavior])
+	_add_to_callbacks(p_behavior)
+
+func _add_to_callbacks(p_behavior: Behavior) -> void:
+	for callback in _callbacks:
+		if p_behavior.has_method(callback):
```

Comment:
If you made the other suggested changes, you would also need to add an `and p_behavior.enabled:` to this condition.

Reply:
what if you enable it afterwards?!

Reply:
^ This is why you would *also* need a setter for the `enabled` property. So that it can, upon being set to a new value, update the Behavior's presence in the Behaviors._callback Dictionary.